container/libcontainer/helpers.go

```diff
@@ -93,14 +99,101 @@ func GetStats(cgroupManager cgroups.Manager, networkInterfaces []string) (*info.
 		}
 		stats.Network.Interfaces[i] = interfaceStats
 	}
+
+	// If we know the pid & we haven't discovered any network interfaces yet
+	// try the network namespace.
+	if pid > 0 && len(stats.Network.Interfaces) == 0 {
+		nsStats, err := networkStatsFromNs(pid)
+		if err != nil {
+			glog.V(2).Infof("Unable to get network stats from pid %d: %v", pid, err)
+		} else {
+			stats.Network.Interfaces = append(stats.Network.Interfaces, nsStats...)
+		}
+	}
+
 	// For backwards compatibility.
-	if len(networkInterfaces) > 0 {
+	if len(stats.Network.Interfaces) > 0 {
 		stats.Network.InterfaceStats = stats.Network.Interfaces[0]
 	}
 
 	return stats, nil
 }
 
+func networkStatsFromNs(pid int) ([]info.InterfaceStats, error) {
+	// Lock the OS Thread so we only change the ns for this thread exclusively
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	stats := []info.InterfaceStats{}
+
+	// Save the current network namespace
+	origns, _ := netns.Get()
+	defer origns.Close()
+
+	// Switch to the pid netns
+	pidns, err := netns.GetFromPid(pid)
+	defer pidns.Close()
+	if err != nil {
+		return stats, nil
+	}
+	netns.Set(pidns)
+
+	// Defer setting back to original ns
+	defer netns.Set(origns)
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return stats, fmt.Errorf("cannot find interfaces: %v", err)
+	}
+
+	// For some reason ioutil.ReadFile(netStatsFile) reads the file in
+	// the default netns when this code is invoked from docker.
+	// Executing "cat <netStatsFile>" works as expected.
+	data, err := exec.Command("cat", netStatsFile).Output()
+	if err != nil {
+		return stats, fmt.Errorf("failure opening %s: %v", netStatsFile, err)
+	}
+
+	for _, iface := range ifaces {
+		if iface.Flags&net.FlagUp != 0 && iface.Flags&net.FlagLoopback == 0 {
+			ifaceStats, err := scanInterfaceStats(string(data), iface.Name)
+			if err != nil {
+				return stats, fmt.Errorf("couldn't read network stats for %s: %v", iface.Name, err)
+			}
+			stats = append(stats, ifaceStats)
+		}
+	}
+
+	return stats, nil
+}
+
+// Borrowed from libnetwork Stats - https://github.com/docker/libnetwork/blob/master/sandbox/interface_linux.go
+// In older kernels (like the one in Centos 6.6 distro) sysctl does not have netns support. Therefore
+// we cannot gather the statistics from /sys/class/net/<dev>/statistics/<counter> files. Per-netns stats
+// are naturally found in /proc/net/dev in kernels which support netns (ifconfig relyes on that).
+const (
+	netStatsFile = "/proc/net/dev"
+	base         = "[  ]*%s:([  ]+[0-9]+){16}"
+)
+
+func scanInterfaceStats(data, ifName string) (info.InterfaceStats, error) {
```

Comment:
How about parsing the entire file once, and index it based on the Interface name?